File: backend/app/models/activity_log.py
from datetime import datetime, timezone
from beanie import Document # pyright: ignore[reportMissingImports]
import pymongo # pyright: ignore[reportMissingImports]
from pymongo import IndexModel, ASCENDING # pyright: ignore[reportMissingImports]
from pydantic import Field
from typing import Annotated

def utc_now():
    return datetime.now(timezone.utc)

class ActivityLog(Document):
    user_id: str
    action: str  # CREATE_ASK, REPLY, LOGIN, etc.
    metadata: Annotated[dict, Field(default_factory=dict)]
    created_at: Annotated[datetime, Field(default_factory=utc_now)]

    class Settings:
        name = "activity_logs"
        indexes = [
            IndexModel([("user_id", ASCENDING)])
        ]

File: frontend/lib/models/activity_log.dart
class ActivityLog {
  final String id;
  final String userId;
  final String action;
  final Map<String, dynamic> metadata;
  final DateTime createdAt;

  ActivityLog({
    required this.id,
    required this.userId,
    required this.action,
    required this.metadata,
    required this.createdAt,
  });

  factory ActivityLog.fromJson(Map<String, dynamic> json) {
    return ActivityLog(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      action: json['action'] as String,
      metadata: json['metadata'] != null ? Map<String, dynamic>.from(json['metadata']) : {},
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
